src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxDocumentConverter.cs

```diff
@@ -0,0 +1,252 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Globalization;
+using Corax;
+using Raven.Client;
+using Raven.Server.Documents.Indexes.Persistence.Corax.WriterScopes;
+using Raven.Server.Json;
+using Raven.Server.Utils;
+using Sparrow.Extensions;
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+using Sparrow.Server;
+using Sparrow.Threading;
+using Voron;
+using Encoding = System.Text.Encoding;
+
+namespace Raven.Server.Documents.Indexes.Persistence.Corax
+{
+    public sealed class CoraxDocumentConverter : ConverterBase
+    {
+        private readonly ByteStringContext _allocator;
+        private readonly Dictionary<Slice, int> _knownFields;
+        private static readonly byte[] _trueLiteral = Encoding.UTF8.GetBytes("true"); 
+        private static readonly byte[] _falseLiteral = Encoding.UTF8.GetBytes("false"); 
+
+        public CoraxDocumentConverter(
+                Index index, 
+                bool indexImplicitNull = false, 
+                bool indexEmptyEntries = true, 
+                string keyFieldName = null, 
+                bool storeValue = false, 
+                string storeValueFieldName = Constants.Documents.Indexing.Fields.ReduceKeyValueFieldName) :
+            base(index, storeValue)
+        {
+            _allocator = new ByteStringContext(SharedMultipleUseFlag.None);
+            _knownFields = GetKnownFields();
+        }
+
+        public Dictionary<Slice, int> GetKnownFields()
+        {
+            if (_knownFields != null)
+                return _knownFields;
+            var knownFields = new Dictionary<Slice, int>();
+            foreach (var field in _fields.Values)
+            {
+                if(field.Name == null) 
+                    continue;
+                
+                Slice.From(_allocator, field.Name, ByteStringType.Immutable, out var value);
+                knownFields.Add(value, field.Id);
+            }
+
+            return knownFields;
+        }
+        
+        public Span<byte> InsertDocumentFields(LazyStringValue key, LazyStringValue sourceDocumentId, object doc, JsonOperationContext indexContext, out LazyStringValue id)
+        {
+            var document = (Document)doc;
+            _allocator.Allocate(document.Data.Size + _fields.Count * 1024, out ByteString buffer);
+            var entryWriter = new IndexEntryWriter(buffer.ToSpan(), _knownFields);
+            id = document.LowerId;
+            bool shouldSkip;
+            
+            
+            //TODO maciej - please look at this.
```

Comment:
I don't understand your intent here?

Reply:
Relevant to this case: https://github.com/ravendb/ravendb/blob/0b77beea631c583fd77b7ead2bf562885126f4e1/src/Corax/IndexEntryBuilder.cs#L150

I'm writing enumerable collection into the buffer one-by-one but I need to keep the length of items that I had (to write them at the end of buffer) - so I'm writing them into the list (list is reused for every Enumerable field per Document, reallocate when needed more space). Basic problem is that I don't know how much memory we need to allocate to keep all lengths. I think this list is the right answer to this problem (for now) but I don't know where should I initialize the list and when to dispose it. (for example to cover all indexing process).